### ISSUE TITLE
[New Edit Mode] Remove edit-related concerns from Browse controllers

### ISSUE THREAD

**Issue by VWoeltjen** [nasa/openmctweb#627]
From #481:

> * Remove edit-related concerns from Browse controllers
  * Specifically, BrowseController and BrowseObjectController. 
  * Ideally, removing the "edit" bundle and leaving the "browse" bundle in place should mean that browsing just works, but that editing is no longer an option
  * Does not necessarily mean these need to be identical to their pre-NEM state, just that NEM-specific concerns should not have a home there. Completely valid to introduce new hooks or extension points to Browse to facilitate NEM, so long as those are logically within the scope of Browse responsibilities.

**Comment by akhenry** [comment]
Totally agree. The question is whether edit 'mode' is a navigation event, or just an object state. Especially in light of the API proposal to make navigation changes URL-driven. If it's a navigation then it implies an `EditController`, if not, then something more subtle is necessary.

My thoughts are that encoding the edit state in the URL implies to me that it's an application-wide state, and that's a bit limiting. I'm inclined to think that it should be an object state, and that a view could potentially contain a mixture of objects in browse and edit states (for example, in sub-object editing). This would be difficult to capture in a URL.

**Comment by VWoeltjen** [comment]
> The question is whether edit 'mode' is a navigation event, or just an object state.

Good question to raise; I definitely share the feeling that this doesn't belong in the URL, at least not as it is currently. I hadn't thought far enough ahead to think of it as "object state" but I think that's totally right - Old Edit Mode was a mode of the _application_, whereas New Edit Mode is a mode of _objects_.

The URL does refer to objects, so I wouldn't say it's wrong to include editing here (although it would make more sense as an `editing=true` query parameter in its current incarnation). To your point, though, that doesn't capture sub-object editing, so my inclination would be to maintain this state separately from the URL, and trigger it from an `editing=true` parameter only if we ever need something like that (which I don't really anticipate.)

Would the `status` capability be useful here? I know we're using the `editing` status for CSS rules currently; could other behavior also detect and react to that?

**Comment by akhenry** [comment]
> Would the status capability be useful here? I know we're using the editing status for CSS rules currently; could other behavior also detect and react to that?

Yes, that's exactly what I was thinking. It's perfect for this purpose, and code can listen for state changes.

**Comment by larkin** [comment]
Editing status is definitely an application status-- while editing is a property of objects, the entire application is in an editing state.  Navigation events are hijacked to ensure changes aren't discarded, almost every component changes (toolbar, main view, inspector, elements pool, tree view, create menu).  

Whenever you start editing, the currently navigated object becomes the edit context for all future work.  For the duration of this sprint, the object you are editing and the object you have navigated to are the same and will not change unless you stop editing and begin editing another object.

So:
* object types must define a main view for themselves (a layout, a plot, etc)
* object types may define an editable main view (to support advanced editing gestures)
* object types may define a toolbar view for themselves during edit
* object types may define a toolbar view for themselves during browse
* object types may define a property sheet view for themselves
* object types may define an editable property sheet view for themselves

"May" means an object can do something, but if it does not do it, then there is some reasonable fallback, e.g. a browse toolbar would default to one with an edit button, the default property sheet could use the object metadata definition (json schema or something similar) to display properties, and an editable property sheet could use the same object metadata definition to render a form.

Now, this may eventually lead to sub object editing when we add a "select" gesture to certain object types.  Initially, layouts would be the only object to implement "select", and select would trigger changes to the toolbar and property sheet views. [1]

In this case, the currently navigated object can now be different from the object you are editing-- e.g. sub object editing a plot in a layout-- but the navigated object would still dictate the "main view" for the remainder of the edit session.  [1]

[1] this is not a feature that is on the table for the current sprint: things could change.

**Comment by akhenry** [comment]
@VWoeltjen Totally extricating edit from browse mode is kinda tricky. The controllers are fine, no problem there, they should not have edit mode code. Problem is with the markup. Because we are not doing a view change for edit mode, there are edit mode markup elements (eg. edit buttons) mixed in with the browse markup. They are generic in the sense that they use mct-toolbar etc as appropriate, but might specify some edit-mode details such as the action category to use for populating the toolbar.

This does not preclude the edit bundle from being disabled with no ill-effects. The edit-specific markup would just be disabled in browse mode.

All style differences are handled via status related CSS.

**Comment by larkin** [comment]
Note: "application status vs object status" -- Not worth debating the nomenclature.  I shouldn't be so absolute; it doesn't really matter.

For upcoming JPL functionality, our minimum support is:
* object types may define an editable property sheet view for themselves

There are improvements, such as:
* object types may define a property sheet view for themselves

But these are not "minimum viable option".

@akhenry is working on this.

**Comment by akhenry** [comment]
Conversation is in danger of drifting out of scope.

> Note: "application status vs object status" -- Not worth debating the nomenclature. I shouldn't be so absolute; it doesn't really matter.

Agreed. I see them as having a conceptual distinction, practically I'm not sure what the actual impact is as although its an object state, the application is distinctly object focused :)

> * object types may define a toolbar view for themselves during edit
> * object types may define a toolbar view for themselves during browse

Maybe? I'd like to understand this a little better. We have a means of customizing toolbar options already via policies, and categories. Is there a need to define a view here?
> * object types may define an editable property sheet view for themselves

Yep. This will be a per-type option defined during registration. I'll ticket this separately
> * object types may define a property sheet view for themselves

Yeeeeaaaaamaybe? Could be a good way of separating concerns, or could just be redundant.

**Comment by larkin** [comment]
The general idea is "if you want to do things the platform way then you use the x interface (policy, actions, whatever), but in case you have something we haven't ever thought of before, you can use arbitrary extension point Y."

The "platform" Property sheets would be some kind of form generation / metadata description system. Ties into the metadata capability. 

Toolbars, well I feel that policy and actions are theoretically reusable, but in practice they are usually single purpose and could be simpler.

**Comment by larkin** [comment]
As far as edit vs browse: if editing is an application state, we could decorate the object provider such that it makes (or attempts to make) all objects editable. And then not have to worry about wrapping capabilities / etc. so objects don't really care that they are being edited.

**Comment by VWoeltjen** [comment]
> @VWoeltjen Totally extricating edit from browse mode is kinda tricky. The controllers are fine, no problem there, they should not have edit mode code. Problem is with the markup. Because we are not doing a view change for edit mode, there are edit mode markup elements (eg. edit buttons) mixed in with the browse markup. They are generic in the sense that they use mct-toolbar etc as appropriate, but might specify some edit-mode details such as the action category to use for populating the toolbar.

> This does not preclude the edit bundle from being disabled with no ill-effects. The edit-specific markup would just be disabled in browse mode.

I see your point, have encountered this difficulty before. You _could_ use an approach like the Time Conductor uses, where it intercepts the element being displayed (by way of a Representer) and tacks on some more elements. This works but is rather hacky, and will likely be sensitive to other markup changes.

I think the better way, going forward, would be to improve our View API (in the context of #463) to include the concept of Regions (places where views can be inserted) that @larkin has proposed. Then, a BrowseObjectView just needs to expose some Regions, and then Edit mode can populate these with the appropriate additional markup. This is consistent with the "add extension points where needed, but avoid mixing concerns" approach.

In light of that, think we can leave the markup end of this refactor until after #463 (perhaps breaking off into another followup issue?)

**Comment by akhenry** [comment]
That sounds reasonable. After posting my previous I hit upon a way of doing this, that I think is somewhere in between the hackiness of the Time Conductor approach, and the full blown View API. Basically the 'browse-object' representation (ie. the 'visualization' of the object itself) in `browse.html` now uses a variable for its key. It defaults to browse-object, but the `EditRepresenter` will listen for changes to edit status and change the variable in-scope to 'edit-object' when the object changes state. This sounds like it might be a step on the road to regions because it's defining a changeable region of the view. This could then be made configurable as a later step.

**Comment by akhenry** [comment]
Linking to https://github.com/nasa/openmctweb/issues/464 which contains notes which may be useful for the in-place refactoring of NEM

**Comment by akhenry** [comment]
There is a branch addressing this in open627_rebase. It is based on the regions functionality introduced in #638 so I will wait until that is finalized and merged to finish this fix, as it will require some refactoring based on code changes from #651

**Comment by charlesh88** [comment]
Testathon 03032016 running on codriver
Saw some significant delays when switching from browse to edit mode, and while editing. Delays seem to correlate to the size of the object's contents: a complex timeline like "RP15 Timeline Rev A" would take 3 to 4 seconds from click to mode switch; a similar delay was observed in selecting a timeline's row while in edit mode. Clicking to edit a simpler layout, like mshirley/Mastcam Images, had a delay of about a second.

**Comment by akhenry** [comment]
@charlesh88 Believe this is part of a broader performance issue with timelines, as reported in #715

**Comment by akhenry** [comment]
Closing this issue as the refactoring has been addressed by #712. The comments from @charlesh88 have been moved to a new issue for performance problems #715
